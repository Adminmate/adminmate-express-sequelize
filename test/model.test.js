// see https://stackoverflow.com/questions/46227783/encoding-not-recognized-in-jest-js
require('mysql2/node_modules/iconv-lite').encodingExists('foo');

// Hide console.log
global.console = {
  log: jest.fn(),
  // Keep native behaviour for other methods
  error: console.error,
  warn: console.warn,
  info: console.info,
  debug: console.debug,
};

const supertest = require('supertest');
const jwt = require('jwt-simple');

// Include the app
const app = require('./app.js');

// Endpoint prefix
const prefix = '/adminmate/api';

// Generate the admin token
const adminToken = jwt.encode({
  exp_date: Date.now() + 1000
}, 'authkey_secret');

// Before all
beforeAll(done => {
  done();
});

// Models properties
describe('Testing POST /api/models/:model', () => {
  it('should return a 403 http response', async () => {
    // Make request
    const response = await supertest(app)
      .post(prefix + '/models/users');

    // Check response
    expect(response.status).toBe(403);
    expect(response.body.code).toBe('not_authorized');
  });

  it('should return a 200 http response', async () => {
    // Make request
    const response = await supertest(app)
      .post(prefix + '/models/users')
      .set('x-access-token', adminToken);

    // Check response
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });
});

// After all
afterAll(done => {
  done();
})