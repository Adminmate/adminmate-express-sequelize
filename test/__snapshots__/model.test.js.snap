// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing POST /api/models/cars With "fields" parameter (name & manufacturer only) 1`] = `
Object {
  "count": 20,
  "data": Array [
    Object {
      "id": 20,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 LM",
    },
    Object {
      "id": 19,
      "manufacturer": "Ferrari",
      "name": "Ferrari 275",
    },
    Object {
      "id": 18,
      "manufacturer": "Ferrari",
      "name": "",
    },
    Object {
      "id": 17,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Lusso",
    },
    Object {
      "id": 16,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GTO",
    },
    Object {
      "id": 15,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT 2+2",
    },
    Object {
      "id": 14,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Berlinetta SWB",
    },
    Object {
      "id": 13,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Coupé",
    },
    Object {
      "id": 12,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 MM",
    },
    Object {
      "id": 11,
      "manufacturer": "Ferrari",
      "name": "Ferrari 212 Inter",
    },
  ],
  "pagination": Object {
    "count": 2,
    "current": 1,
  },
}
`;

exports[`Testing POST /api/models/cars With "page" parameter set to 2 1`] = `
Object {
  "count": 20,
  "data": Array [
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 10,
      "manufacturer": "Ferrari",
      "name": "Ferrari 166 MM",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 10,
        "type": "ref",
      },
      "year": 1948,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 9,
      "manufacturer": "Porsche",
      "name": "Porsche 924",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 9,
        "type": "ref",
      },
      "year": 1976,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 8,
      "manufacturer": "Porsche",
      "name": "Porsche 917",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 8,
        "type": "ref",
      },
      "year": 1969,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 7,
      "manufacturer": "Porsche",
      "name": "Porsche 914",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": 7,
        "type": "ref",
      },
      "year": 1969,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 6,
      "manufacturer": "Porsche",
      "name": "Porsche 908",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 6,
        "type": "ref",
      },
      "year": 1968,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 5,
      "manufacturer": "Porsche",
      "name": "Porsche 906",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": 5,
        "type": "ref",
      },
      "year": 1966,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 4,
      "manufacturer": "Porsche",
      "name": "Porsche 904",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 4,
        "type": "ref",
      },
      "year": 1964,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 3,
      "manufacturer": "Porsche",
      "name": "Porsche 911",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": 3,
        "type": "ref",
      },
      "year": 1963,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 2,
      "manufacturer": "Porsche",
      "name": "Porsche 550",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 2,
        "type": "ref",
      },
      "year": 1953,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 1,
      "manufacturer": "Porsche",
      "name": "Porsche 356",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": 1,
        "type": "ref",
      },
      "year": 1948,
    },
  ],
  "pagination": Object {
    "count": 2,
    "current": 2,
  },
}
`;

exports[`Testing POST /api/models/cars With "refFields" parameter 1`] = `
Object {
  "count": 20,
  "data": Array [
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 20,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 LM",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": "Maria Doe",
        "type": "ref",
      },
      "year": 1964,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 19,
      "manufacturer": "Ferrari",
      "name": "Ferrari 275",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": "Maria Doe",
        "type": "ref",
      },
      "year": 1964,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 18,
      "manufacturer": "Ferrari",
      "name": "",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": "Maria Doe",
        "type": "ref",
      },
      "year": 1969,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 17,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Lusso",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": "John Doe",
        "type": "ref",
      },
      "year": 1962,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 16,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GTO",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": "Maria Doe",
        "type": "ref",
      },
      "year": 1962,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 15,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT 2+2",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": "John Doe",
        "type": "ref",
      },
      "year": 1960,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 14,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Berlinetta SWB",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": "Maria Doe",
        "type": "ref",
      },
      "year": 1959,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 13,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Coupé",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": "John Doe",
        "type": "ref",
      },
      "year": 1954,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 12,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 MM",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": "Maria Doe",
        "type": "ref",
      },
      "year": 1952,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 11,
      "manufacturer": "Ferrari",
      "name": "Ferrari 212 Inter",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": "John Doe",
        "type": "ref",
      },
      "year": 1951,
    },
  ],
  "pagination": Object {
    "count": 2,
    "current": 1,
  },
}
`;

exports[`Testing POST /api/models/cars With a "order" parameter 1`] = `
Object {
  "count": 3,
  "data": Array [
    Object {
      "id": 3,
      "name": "Porsche 911",
    },
    Object {
      "id": 7,
      "name": "Porsche 914",
    },
    Object {
      "id": 8,
      "name": "Porsche 917",
    },
  ],
  "pagination": Object {
    "count": 1,
    "current": 1,
  },
}
`;

exports[`Testing POST /api/models/cars With a "search" parameter 1`] = `
Object {
  "count": 3,
  "data": Array [
    Object {
      "id": 8,
      "name": "Porsche 917",
    },
    Object {
      "id": 7,
      "name": "Porsche 914",
    },
    Object {
      "id": 3,
      "name": "Porsche 911",
    },
  ],
  "pagination": Object {
    "count": 1,
    "current": 1,
  },
}
`;

exports[`Testing POST /api/models/cars With access token 1`] = `
Object {
  "count": 20,
  "data": Array [
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 20,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 LM",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 20,
        "type": "ref",
      },
      "year": 1964,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 19,
      "manufacturer": "Ferrari",
      "name": "Ferrari 275",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 19,
        "type": "ref",
      },
      "year": 1964,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 18,
      "manufacturer": "Ferrari",
      "name": "",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 18,
        "type": "ref",
      },
      "year": 1969,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 17,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Lusso",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": 17,
        "type": "ref",
      },
      "year": 1962,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 16,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GTO",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 16,
        "type": "ref",
      },
      "year": 1962,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 15,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT 2+2",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": 15,
        "type": "ref",
      },
      "year": 1960,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 14,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Berlinetta SWB",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 14,
        "type": "ref",
      },
      "year": 1959,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 13,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 GT Coupé",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": 13,
        "type": "ref",
      },
      "year": 1954,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 12,
      "manufacturer": "Ferrari",
      "name": "Ferrari 250 MM",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 2,
        "label": 12,
        "type": "ref",
      },
      "year": 1952,
    },
    Object {
      "createdAt": "2021-03-13T23:00:00.000Z",
      "id": 11,
      "manufacturer": "Ferrari",
      "name": "Ferrari 212 Inter",
      "updatedAt": "2021-03-13T23:00:00.000Z",
      "userId": Object {
        "id": 1,
        "label": 11,
        "type": "ref",
      },
      "year": 1951,
    },
  ],
  "pagination": Object {
    "count": 2,
    "current": 1,
  },
}
`;

exports[`Testing POST /api/models/users With access token 1`] = `
Object {
  "count": 2,
  "data": Array [
    Object {
      "birthdate": "2021-03-13T22:49:15.000Z",
      "createdAt": "2021-03-13T22:49:15.000Z",
      "firstname": "Maria",
      "id": 2,
      "lastname": "Doe",
      "rating": 5,
      "updatedAt": "2021-03-13T22:49:15.000Z",
    },
    Object {
      "birthdate": "2021-03-13T22:49:15.000Z",
      "createdAt": "2021-03-13T22:49:15.000Z",
      "firstname": "John",
      "id": 1,
      "lastname": "Doe",
      "rating": 4,
      "updatedAt": "2021-03-13T22:49:15.000Z",
    },
  ],
  "pagination": Object {
    "count": 1,
    "current": 1,
  },
}
`;
